### ISSUE TITLE
Enhancement: Improve Neural Network Accuracy

### ISSUE THREAD

**Issue by mranv** [09Kanika/Neural-Network-from-scratch#1]
### Description
**Issue:**
The current neural network implementation achieves a 94% accuracy on the provided dataset. However, there is room for improvement to further increase the accuracy and overall performance of the model.

**Proposed Enhancements:**

1. **Weight Initialization:**
   - Modify the weight initialization to smaller values to avoid issues with exploding/vanishing gradients.
   ```python
   self.weights = 0.01 * np.random.randn(n_inputs, n_neurons)
   ```

2. **Learning Rate Adjustment:**
   - Implement a dynamic learning rate and consider using a learning rate scheduler for better convergence.
   ```python
   learning_rate = 0.01  # Adjust as needed
   ```

3. **Optimizers:**
   - Integrate more sophisticated optimizers like Adam or SGD with momentum to improve the training process.
   ```python
   class Optimizer_SGD:
       def __init__(self, learning_rate=1.0, decay=0.0, momentum=0.0):
           # Initialization code...

       def update_params(self, layer):
           # Update logic...
   ```

4. **Regularization:**
   - Implement regularization techniques such as dropout and L2 regularization to prevent overfitting.
   ```python
   class Layer_Dense:
       def __init__(self, n_inputs, n_neurons, l2_reg_lambda=0.01):
           # Initialization with regularization
   ```

5. **Early Stopping:**
   - Add early stopping to terminate training when validation loss stops improving to avoid overfitting.

6. **Data Normalization:**
   - Normalize input data to improve training stability.
   ```python
   x -= np.mean(x, axis=0)
   x /= np.std(x, axis=0)
   ```

7. **Backpropagation and Training Loop:**
   - Ensure proper implementation of backpropagation to update weights based on gradients, and refine the training loop.
   ```python
   class Layer_Dense:
       def backward(self, dvalues):
           # Backpropagation logic...
   ```

**Example Code:**
Here is an example of the refined training loop:
```python
for epoch in range(1001):
    dense1.forward(x)
    activation1.forward(dense1.output)
    dense2.forward(activation1.output)
    activation2.forward(dense2.output)

    loss = loss_function.calculate(activation2.output, y)
    
    predictions = np.argmax(activation2.output, axis=1)
    accuracy = np.mean(predictions == y)
    
    if epoch % 50 == 0:
        print(f'epoch: {epoch}, loss: {loss}, accuracy: {accuracy}')
    
    loss_function.backward(activation2.output, y)
    dense2.backward(loss_function.dinputs)
    activation1.backward(dense2.dinputs)
    dense1.backward(activation1.dinputs)
    
    optimizer.pre_update_params()
    optimizer.update_params(dense1)
    optimizer.update_params(dense2)
    optimizer.post_update_params()
```

By implementing these enhancements, we can potentially increase the accuracy and robustness of the neural network model.
